File: cli/internal/git/git.go
package git

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// GetCurrentBranch returns the current git branch name
func GetCurrentBranch(dir string) (string, error) {
	// First try reading .git/HEAD directly (faster, no exec)
	headPath := filepath.Join(dir, ".git", "HEAD")
	data, err := os.ReadFile(headPath)
	if err == nil {
		content := strings.TrimSpace(string(data))
		// Format: "ref: refs/heads/branch-name"
		if strings.HasPrefix(content, "ref: refs/heads/") {
			return strings.TrimPrefix(content, "ref: refs/heads/"), nil
		}
		// Detached HEAD - return the SHA
		return content[:7], nil
	}

	// Fall back to git command
	cmd := exec.Command("git", "rev-parse", "--abbrev-ref", "HEAD")
	cmd.Dir = dir
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}

	return strings.TrimSpace(string(output)), nil
}

// IsGitRepo checks if the directory is inside a git repository
func IsGitRepo(dir string) bool {
	gitDir := filepath.Join(dir, ".git")
	info, err := os.Stat(gitDir)
	if err != nil {
		return false
	}
	return info.IsDir()
}

// GetRepoRoot returns the root directory of the git repository
func GetRepoRoot(dir string) (string, error) {
	cmd := exec.Command("git", "rev-parse", "--show-toplevel")
	cmd.Dir = dir
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(output)), nil
}

// GetRemoteURL returns the origin remote URL
func GetRemoteURL(dir string) (string, error) {
	cmd := exec.Command("git", "remote", "get-url", "origin")
	cmd.Dir = dir
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(output)), nil
}

// HasUncommittedChanges checks if there are uncommitted changes
func HasUncommittedChanges(dir string) (bool, error) {
	cmd := exec.Command("git", "status", "--porcelain")
	cmd.Dir = dir
	output, err := cmd.Output()
	if err != nil {
		return false, err
	}
	return len(strings.TrimSpace(string(output))) > 0, nil
}

// GetHeadCommit returns the current HEAD commit SHA
func GetHeadCommit(dir string) (string, error) {
	cmd := exec.Command("git", "rev-parse", "HEAD")
	cmd.Dir = dir
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(output)), nil
}

File: cli/src/lib/git.ts
/**
 * Git utilities
 */

import { execSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";

/**
 * Get the current git branch name
 */
export function getCurrentBranch(dir: string): string | null {
  // Try reading .git/HEAD directly first (faster)
  const headPath = join(dir, ".git", "HEAD");

  if (existsSync(headPath)) {
    try {
      const content = readFileSync(headPath, "utf-8").trim();

      // Format: "ref: refs/heads/branch-name"
      if (content.startsWith("ref: refs/heads/")) {
        return content.replace("ref: refs/heads/", "");
      }

      // Detached HEAD - return short SHA
      return content.slice(0, 7);
    } catch {
      // Fall through to git command
    }
  }

  // Fall back to git command
  try {
    const branch = execSync("git rev-parse --abbrev-ref HEAD", {
      cwd: dir,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    }).trim();

    return branch;
  } catch {
    return null;
  }
}

/**
 * Check if directory is a git repository
 */
export function isGitRepo(dir: string): boolean {
  const gitDir = join(dir, ".git");
  return existsSync(gitDir);
}

/**
 * Get the git repository root
 */
export function getRepoRoot(dir: string): string | null {
  try {
    const root = execSync("git rev-parse --show-toplevel", {
      cwd: dir,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    }).trim();

    return root;
  } catch {
    return null;
  }
}

/**
 * Get the remote origin URL
 */
export function getRemoteUrl(dir: string): string | null {
  try {
    const url = execSync("git remote get-url origin", {
      cwd: dir,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    }).trim();

    return url;
  } catch {
    return null;
  }
}

/**
 * Check for uncommitted changes
 */
export function hasUncommittedChanges(dir: string): boolean {
  try {
    const status = execSync("git status --porcelain", {
      cwd: dir,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    }).trim();

    return status.length > 0;
  } catch {
    return false;
  }
}

/**
 * Get the current HEAD commit SHA
 */
export function getHeadCommit(dir: string): string | null {
  try {
    const sha = execSync("git rev-parse HEAD", {
      cwd: dir,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    }).trim();

    return sha;
  } catch {
    return null;
  }
}
